### ISSUE TITLE
unpaper removes half page at 600dpi

### ISSUE THREAD

**Issue by milahu** [unpaper/unpaper#230]
this works as expected with 300dpi, but fails with 600dpi

this works on normal pages full of text, but fails with an almost empty page

command

```
unpaper -vv --dpi 600 --input-pages 1 --output-pages 1 --layout single \
  --no-multi-pages --no-deskew \
  page.2024-07-16.10-45-28.0005.tiff unpaper/page.2024-07-16.10-45-28.0005.tiff
```

i had to add `--no-deskew` because the output was worse than the input
the other options are an attempt to fix the "removes half page" issue, but no effect

i need unpaper only to remove the black frame around the page
(i know its overkill and i should use opencv, but well...)

<details>
<summary>
input: Für meinen Vater
</summary>

translation: `Für meinen Vater` means `For my father`

![page 2024-07-16 10-45-28 0005 tiff](https://github.com/user-attachments/assets/682bbbeb-b2cc-40d2-b246-00ee8274273c)

</details>

<details>
<summary>
output: Für meinen
</summary>

![page 2024-07-16 10-45-28 0005 tiff](https://github.com/user-attachments/assets/22560824-1fd0-4072-915c-e8164a5059b7)

</details>

<details>
<summary>
log output
</summary>

```
Input #0, png_pipe, from 'out/page.2024-07-16.10-45-28.0005.tiff.png':
  Duration: N/A, bitrate: N/A
  Stream #0:0: Video: png, gray(pc), 1444x2359, 25 fps, 25 tbr, 25 tbn
Output #0, image2, to 'out/page.2024-07-16.10-45-28.0005.tiff.png.unpaper.png':
  Stream #0:0: Unknown: none
[image2 @ 0x176e540] Encoder did not produce proper pts, making some up.
[image2 @ 0x176e540] The specified filename 'out/page.2024-07-16.10-45-28.0005.tiff.png.unpaper.png' does not contain an image sequence pattern or a pattern is invalid.
[image2 @ 0x176e540] Use a pattern such as %03d for an image sequence or use the -update option (with -frames:v 1 if needed) to write a single image.
unpaper 7.0.0
License GPLv2: GNU GPL version 2.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

-------------------------------------------------------------------------------
Processing sheet: out/page.2024-07-16.10-45-28.0005.tiff.png -> out/page.2024-07-16.10-45-28.0005.tiff.png.unpaper.png
loading file out/page.2024-07-16.10-45-28.0005.tiff.png.
layout: single
blackfilter-scan-direction: [vertical,horizontal]
blackfilter-scan-size: [20,20]
blackfilter-scan-depth: [500,500]
blackfilter-scan-step: [5,5]
blackfilter-scan-threshold: 0.950000
blackfilter-intensity: 20
noisefilter-intensity: 4
blurfilter-size: [100,100]
blurfilter-step: [50,50]
blurfilter-intensity: 0.010000
grayfilter-size: [50,50]
grayfilter-step: [20,20]
grayfilter-threshold: 0.500000
mask points: 
mask-scan-direction: [horizontal]
mask-scan-size: [50,50]
mask-scan-depth: [-1,-1]
mask-scan-step: [5,5]
mask-scan-threshold: [0.100000,0.100000]
mask-scan-minimum: [100,100]
mask-scan-maximum: [-1,-1]
mask-color: 16777215
deskew-scan DISABLED for all sheets.
border-scan-direction: [vertical]
border-scan-size: [5,5]
border-scan-step: [5,5]
border-scan-threshold: [5,5]
border-align: []
border-margin: [0,0]
white-threshold: 0.900000
black-threshold: 0.330000
sheet-background: white ffffff
dpi: 300
input-files per sheet: 1
output-files per sheet: 1
input-file-sequence:  out/page.2024-07-16.10-45-28.0005.tiff.png
output-file-sequence: out/page.2024-07-16.10-45-28.0005.tiff.png.unpaper.png

input-file for sheet 1: out/page.2024-07-16.10-45-28.0005.tiff.png
output-file for sheet 1: out/page.2024-07-16.10-45-28.0005.tiff.png.unpaper.png
sheet size: 1444x2359
...
black-area flood-fill: [1420,0,1439,499]
noise-filter ... deleted 82 clusters.
blur-filter... deleted 22 pixels.
auto-masking (722,1179): 252,0,1002,2358
gray-filter... deleted 20544408 pixels.
+ deskewing DISABLED for sheet 1
auto-masking (722,1179): 397,0,987,2358
centering mask [397,0,987,2358] (722,1179): 30, 0
border detected: (0,765,1,1541) in [0,0,1443,2358]
aligning mask [0,765,1442,817] (0,1152): 0, 387
writing output.
saving file out/page.2024-07-16.10-45-28.0005.tiff.png.unpaper.png.
```

</details>

similar issue:
unpaper removes left part of page at 600dpi

<details>
<summary>
input
</summary>

![page 2024-07-18 16-17-44 0204 tiff](https://github.com/user-attachments/assets/a0e7e5f6-4e03-4b69-a55b-21a229e110e1)

</details>

<details>
<summary>
output
</summary>

![page 2024-07-18 16-17-44 0204 tiff](https://github.com/user-attachments/assets/9d7086df-f70b-468f-ac66-d98678eeb5dd)

</details>

i guess that unpaper is irritated by the fact
that i have already removed the black shadow/frame/border on the inner book side

i remove the inner book edge manually to center the text block

the original scanned pages look like this

<details>
<summary>
scanned page
</summary>

the vertical grey line is noise from my ADF scanner ([Brother ADS-3000N](https://geizhals.de/brother-imagecenter-ads-3000n-ads3000nux1-a1375507.html))
i remove it with `color_levels(img, 130, 200)` in preprocess.py

![page 2024-07-16 10-45-28 0005 tiff](https://github.com/user-attachments/assets/0bb83af8-9466-43ae-97c1-de6517e9b2f1)

</details>

im using this script to preprocess the scanned pages

<details>
<summary>
preprocess.py
</summary>

```py
#!/usr/bin/env python3

import cv2
import numpy as np
from skimage.filters import threshold_otsu
from matplotlib import pyplot as plt
from sklearn.cluster import KMeans
from itertools import combinations
from collections import defaultdict



import os
import sys
import math
import imutils
import time



def color_levels(img, low, high):

    # color levels: increase contrast
    # needed to ignore the extra black background created by rotating the image
    # https://stackoverflow.com/questions/26912869/color-levels-in-opencv

    low_in = low
    high_in = high

    low_out = 0 * 255
    high_out = 1 * 255

    if isinstance(img[0][0], (np.uint8, int)):
        num_channels = 1
    else:
        num_channels = len(img[0][0])

    inBlack  = np.array(num_channels * [low_in], dtype=np.float32)
    inWhite  = np.array(num_channels * [high_in], dtype=np.float32)
    inGamma  = np.array(num_channels * [1.0], dtype=np.float32)
    outBlack = np.array(num_channels * [low_out], dtype=np.float32)
    outWhite = np.array(num_channels * [high_out], dtype=np.float32)

    img = np.clip( (img - inBlack) / (inWhite - inBlack), 0, 255)
    img = ( img ** (1/inGamma) ) *  (outWhite - outBlack) + outBlack
    img = np.clip( img, 0, 255).astype(np.uint8)
    """
    # https://docs.opencv.org/3.4/d7/d4d/tutorial_py_thresholding.html # otsu automatic threshold
    # https://stackoverflow.com/questions/20049995/thresholding-of-a-grayscale-image-in-a-range
    low, high = 0.5*255, 0.6*255
    low, high = 0.25*255, 0.75*255
    low, high = 0.5*255, 0.7*255 # tested with gimp. too dark
    low, high = 0.4*255, 0.8*255
    #img = cv2.inRange(img, low, high)
    gray = cv2.inRange(gray, low, high)
    num_channels = 3
    img = cv2.inRange(img, tuple([low] * num_channels), tuple([high] * num_channels))
    """

    return img



def find_edge(img, side="L"):

    debug = True

    # https://stackoverflow.com/questions/45322630/how-to-detect-lines-in-opencv

    H = len(img)
    W = len(img[0])
    # ignore 100px top of image
    y = 100
    # crop 100px band to measure angle of edge
    w = 100
    if side == "L":
        x = 0
    else:
        x = W - w
    #h = H
    # ignore 100px bottom of image
    #h = 4742 - 200 # 600dpi # why 4742
    h = H - 200 # 600dpi

    # crop image
    img = img[y:y+h, x:x+w]

    #img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # no. levels (0.5, 0.7) leave scanner artifacts
    # light-gray vertical lines
    # which are then produce a false-positive match as page edge
    #img = color_levels(img, 0.5 * 255, 0.7 * 255)
    img = color_levels(img, 0.6 * 255, 0.8 * 255)

    #gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    gray = img

    if debug:
        path = f"find_edge.{time.time()}.levels.tiff"
        print("writing", path)
        cv2.imwrite(path, img)

    kernel_size = 5
    #kernel_size = 3 # 600dpi

    blur_gray = cv2.GaussianBlur(gray,(kernel_size, kernel_size),0)

    # Second, process edge detection use Canny.
    low_threshold = 50
    high_threshold = 150
    edges = cv2.Canny(blur_gray, low_threshold, high_threshold)
    #print("edges", edges)

    # Then, use HoughLinesP to get the lines. You can adjust the parameters for better performance.

    # distance resolution in pixels of the Hough grid
    rho = 1

    # angular resolution in radians of the Hough grid
    theta = np.pi / 180

    # minimum number of votes (intersections in Hough grid cell)
    threshold = 15

    # minimum number of pixels making up a line
    min_line_length = 50
    #min_line_length = round(0.8 * H)
    #min_line_length = round(0.5 * H)

    # maximum gap in pixels between connectable line segments
    max_line_gap = 20

    # create a blank to draw lines on
    line_image = np.copy(img) * 0

    # Run Hough on edge detected image
    # Output "lines" is an array containing endpoints of detected line segments
    lines = cv2.HoughLinesP(edges, rho, theta, threshold, np.array([]),
                        min_line_length, max_line_gap)

    if lines is None:
        lines = []

    xmin = W
    ymin = H
    xmax = 0
    ymax = 0

    points = []

    for line in lines:
        for x1,y1,x2,y2 in line:
            #print("line", x1,y1,x2,y2)
            angle = math.degrees(math.atan2(x2 - x1, y2 - y1))
            if angle < 170 or 190 < angle:
                continue
            #if 170 < angle and angle < 190:
            #print("angle", angle)
            cv2.line(line_image,(x1,y1),(x2,y2),(255,0,0),5)
            points.append((x1, y1))
            points.append((x2, y2))

    # sort points by y
    points.sort(key=lambda p: p[1])

    if len(points) < 2:
        return None

    max_line = points[0][0], points[0][1], points[-1][0], points[-1][1]

    print("max_line", max_line)
    x1,y1,x2,y2 = max_line
    edge_angle = math.degrees(math.atan2(x2 - x1, y2 - y1))
    print("edge_angle", edge_angle)

    # Finally, draw the lines on your srcImage.
    # Draw the lines on the  image
    lines_edges = cv2.addWeighted(img, 0.8, line_image, 1, 0)

    if debug:
        path = f"find_edge.{time.time()}.out.tiff"
        print("writing", path)
        cv2.imwrite(path, lines_edges)

    return x1, y1, x2, y2, edge_angle



def process_image(path, pageno):

    out_path = os.path.join(
        os.path.dirname(path),
        "out",
        os.path.basename(path),
    )

    if os.path.exists(out_path):
        print("keeping", out_path)
        return

    os.makedirs(os.path.dirname(out_path), exist_ok=True)

    # First, get the gray image and process GaussianBlur.

    img = cv2.imread(path, cv2.IMREAD_GRAYSCALE)

    side = "L" if pageno % 2 == 0 else "R"

    print("path", path, "pageno", pageno, "side", side)

    x1, y1, x2, y2, edge_angle = find_edge(img, side)
    print("image edge", x1, y1, x2, y2, edge_angle)

    img = imutils.rotate_bound(img, edge_angle)



    # crop page to 2836 x 4700
    # so the text block is in center of page
    # find vertical edge in rotated image
    x1, y1, x2, y2, edge_angle = find_edge(img, side)
    #print("rotated image edge", x1, y1, x2, y2, edge_angle)

    keep_straight_edges = True

    x_edge = round((x1 + x2) / 2)
    #x_edge = max(x1, x2)

    H = len(img)
    W = len(img[0])

    if keep_straight_edges:
        w = 2836 + x_edge
    else:
        w = 2836

    if keep_straight_edges:
        h = H
    else:
        h = 4700

    # always keep top of image
    # nothing to crop here (scanner bug?)
    y = 0

    if side == "L":
        if keep_straight_edges:
            x = 0
        else:
            x = x_edge

    elif side == "R":
        if keep_straight_edges:
            x = W - w
        else:
            # revert cropping in find_edge
            x_edge = W - 100 + x_edge
            x = x_edge - w

    # crop image
    img = img[y:y+h, x:x+w]

    # color levels
    # >>> 130 / 255
    # 0.5098039215686274
    # >>> 200 / 255
    # 0.7843137254901961
    img = color_levels(img, 130, 200)

    # remove top black bar from rotating
    bar_height = 0
    for y in range(20):
        #print("img[y][0]", repr(img[y][0]), img[y][0])
        if img[y][0] == 0: # left side
            bar_height += 1
        elif img[y][-1] == 0: # right side
            bar_height += 1
        else:
            break
    if bar_height:
        print("removing top bar", bar_height)
        img = img[bar_height:]

    # remove bottom noise from scanner imprecisions
    bar_height = 10
    if bar_height:
        print("removing bottom bar", bar_height)
        img = img[:(-1 * bar_height)]

    print("writing", out_path)
    cv2.imwrite(out_path, img)



for path in sys.argv[1:]:
    pageno = int(os.path.basename(path).split(".")[-2])
    process_image(path, pageno)
```

</details>

**Comment by mcorrigan** [comment]
I know this is older, but I ran into this as well. Turns out it has something to do with the mask I believe. When I add a mask-scan-threshold value (over 2 in my specific case) then it won't crop part of the page off. Wish I understood it better so I could really dial it in, but for now this is all I have discovered.
